### DISCUSSION TITLE
Anomalies in skymap

### DISCUSSION THREAD

**Discussion by CassandraAuri** [mshumko/asilib#25] [General]

**Comment by mshumko** [comment]
@CassandraAuri I'm unsure on what happened to your original post but I've been thinking of how to address this for a while and I welcome your solutions.

I suspect that the pokey pixel at the edge has to do with how the NaN (lat, lon) points are collapsed to the nearest valid point, row by now. This works for ASIs whose FOV fully projects onto the sensor (e.g., THEMIS) but now for ASIs whose FOV is larger than the sensor (TREx-RGB).

I've tried two solutions to this:
1. [Changing the algorithm for removing NaNs in the (lat, lon) grid from the row-by-row approach to polar coordinate approach.](https://github.com/mshumko/asilib/blob/3cf5143807cb32bfcdc4d5704fd679e53b2ffbdd/asilib/imager.py#L1795-L1893) I almost have this implemented, but there is a bug that I have not had the chance to debug.
2. [And a while back I've tried to use matpltolib.patches.Polygon](https://github.com/mshumko/asilib/blob/a79f19f8b6870a8c73f502e853a334f5c89f6bb2/asilib/plot/plot_map.py#L113-L138) to plot each pixel individually.

The second method should work, but I abandoned it because it was much slower than pcolormesh. Another solution that I've considered is to move the code to remove NaN values into `asilib.asi.trex_rgb()` and add the modified (lat, lon) arrays into the `skymap` dictionary. Those arrays can be called, for example `lat_valid` and `lon_valid`. Then inside `_pcolormesh_nan`, if it finds those arrays it will plot using those instead of the (lat, lon) grid with NaNs in them.

**Reply by CassandraAuri** [reply]
Hey Mike, some crazy stuff has come up and I am super busy all of a sudden so I will respond. However you are free to look at the jupyter notebook here as it has the testing I've done. I actually messed up and forgot about the IDL -> Python interpretting so after that I was able to atleast reproduce asilib results, however, the trex rgb skymaps are completely wacky and I don't know if I believe them as on the event I selected and am interested in their calibration looks off... I am trying to contact Eric about this but hes gone AWOL so I may email Megan as it seems she made the skymap for the imager.

Difference between ASILIB and straight from skymap file when using IDL -> Python interpreting
![image](https://github.com/user-attachments/assets/b6ca5765-214a-4b7c-bd5d-a261257292a8)

Weird TREX-rgb skymap but I have more examples 
![image](https://github.com/user-attachments/assets/e3ad898a-636a-43d4-b0bd-6d0b2393821b)


https://github.com/CassandraAuri/Physics_Work/blob/Deployment-version/Aurora_Work/read_themis_data.ipynb

I am planning to update the main post when my workload dies down with proper results.

**Comment by mshumko** [comment]
@CassandraAuri thanks for the update and I hope that your workload abates soon. I will take a closer look at this in the next two weeks. Do you have a reference mapped image to use as the ground truth? Otherwise I can dig into U. Calgary's crib sheets.

**Reply by CassandraAuri** [reply]
Workload was due to an essay for an english class I am doing over the summer to get over with. Was due sunday. Ian was also only here a week so lots of meetings to update. Workload is over now. I don't have an absolute truth because I dont have IDL installed but maybe its worth installing that and looking over to see what the IDL version should look like. It could be the IDL -> python is acting up.

**Comment by CassandraAuri** [comment]
Hey @mshumko Played around more and the issue is definitely on our end. Using the trex-rgb mosiac crib sheet, I was able to produce this beautiful plot that shows the aurora very well

![image](https://github.com/user-attachments/assets/0bcc9113-983e-4958-a02c-739ee7c27f07)

however, when doing the same for asilib

![image](https://github.com/user-attachments/assets/07726c63-fcd8-442f-992e-bbe46769000b)

I am going to keep digging

**Reply by CassandraAuri** [reply]
Maybe I'm crazy but it looks like the aurora is at a different angle, the crib sheet one is angled, where the asilib is flat. I think the pokies are actually consisent between the two images, but its not removing the nans correctly in the top right. The crib one seems to be brighter and I argue better as its limiting the magnitude to be [10,110] whereas I can reproduce the asilib color by setting the bounds to [0,256]

**Reply by mshumko** [reply]
I see. I agree that there are three things to address:

1. That spike. I need to track down what pixel that is, and where in the code it gets adjusted (or not adjusted)
2. A relative rotation. Maybe it is due to a different map projection? Or maybe the skymaps were updated? You test this by deleting skymaps from both codes and rerunning your code to redownload the latest skymaps
3. I think that Calgary applies a color adjustment that we can consider doing ourselves

![asilib_vs_calgary_skymaps](https://github.com/user-attachments/assets/deaedf28-3f48-4986-a8e1-8ccd094eaa89)

**Reply by CassandraAuri** [reply]
I just finished implementing a normalization to rgb pr inc, I am going to to then try to get a rid on the calgary map to overplot.
